docs/design-documents/platform/pin-names/pin_names.md

```diff
@@ -0,0 +1,198 @@
+
+# Table of contents
+
+1. [Table of contents](#table-of-contents).
+    * [Revision history](#revision-history).
+1. [Introduction](#introduction).
+    1. [Overview and background](#overview-and-background).
+    1. [Requirements and assumptions](#requirements-and-assumptions).
+1. [Detailed design](#detailed-design).
+    1. [Arduino Uno (Rev 3) Header Pins](#arduino-uno-rev-3-header-pins).
+        1. [I2C Definition for Arduino Uno (Rev3) Pins](#i2c-definition-for-arduino-uno-rev3-pins)
+        1. [SPI Definition for Arduino Uno (Rev3) Pins](#spi-definition-for-arduino-uno-rev3-pins)
+    1. [Custom/Vendor Header Pins](#custom/vendor-header-pins).
+        1. [Custom/Vendor Connector I2C Definition](#custom/vendor-connector-i2c-definition).
+        1. [Custom/Vendor Connector SPI Definition](#custom/vendor-connector-spi-definition).
+    1. [Board Components Pins](#board-components-pins).
+        1. [LED Definition](#led-definition)
+        1. [BUTTON Definition](#button-definition)
+1. [Other information](#other-information).
+    1. [Avoid Unnecessary Definitions](#avoid-unnecessary-definitions).
+
+
+### Revision history
+
+1.0 - Initial revision - Malavika Sajikumar - 3/5/2020.  
+This document was written for MBED OS version 5.15 (released on 11th of February 2020).
+
+# Introduction
+
+### Overview and background
+
+Mbed is designed so that application code written in the platform is portable across different Mbed supported boards with the same hardware capabilities or interfaces. However, the code, in most cases, is not truly portable due to the differences in pin name definitions for the same kind of interfaces across different boards. 
+
+This document provides a standard for defining pin names for the commonly used Mbed interfaces like I2C, SPI, etc. and board components like buttons and LEDs in a board support package.
+
+### Requirements and assumptions
+
+This document mainly applies to the pin standards required for Arduino Uno header which is shared among multiple Mbed enabled boards. The Arduino Uno header has been stable since 2012 at its current revision, which is the Arduino Uno Rev3. All design choices in this document for the Arduino Uno header are based on the Arduino Uno Rev3 header implementation.
+
+# Detailed design
+
+To achieve meaningful portability of application code across various Mbed enabled boards, the pin names of commonly used interfaces and board components should be common across these boards. This document describes a set of rules on how to name these pins in the board support package, specifically in PinNames.h file of the board.
+
+### Arduino Uno (Rev3) Header Pins
+
+All Arduino Uno (Rev3) form factor Mbed controller boards should define D0-D15 & A0-A5 pins as a default standard.
+
+    // Arduino Uno (Rev3) header pin connection naming  
+    // Px_xx relates to the processor pin connected to the Arduino Uno (Rev3) header pin
+    
+    D0 = Px_xx,  // UART RX / GPIO 0  
```

Comment:
What if some of these are not available? I recall use cases where some were just connected  to VCC or similar. This probably is more connected to below (for instance not always i2c is polluted, it could be a case).

Reply:
I think you'll have to define them as `NC`, otherwise stuff won't compile - if you've got `TARGET_FF_ARDUINO` on, existing code will need that.

Reply:
That is my understanding as well, but should be noted here as well.

Reply:
@0xc0170 @kjbracey-arm How much flexibility can pins of Arduino form factor have for a board to still be tagged as TARGET_FF_ARDUINO? If certain pins are not connected or have multiple functionality, how can code portability be achieved?

Reply:
But if the for some reason a pin is not defined for the board (whether in Arduino header or just a BUTTONX) does it make sense for the code to compile? I would rather said it should fail rather than pretend to work...

Reply:
> If certain pins are not connected or have multiple functionality, how can code portability be achieved?

In practice, if the Arduino connector is not fully independent of the rest of the board, as is quite common, then pin conflicts can happen. How much it matters will depend on which pins you're using. Many things wanting to use the connector only want a couple of pins, so good chance of working, even if it's not a 100%-"intact" connector.

Some drivers are capable of surviving with some pins unconnected, and they would be looking for that with `!= NC` tests.

> I would rather said it should fail rather than pretend to work...

Most uses of `NC` lead to either an MBED_ERROR or MBED_ASSERT (I forget which), so it's not subtle. But certainly harder to track than a compile error.

> does it make sense for the code to compile?

It does if the code is capable of being autoconfiguring in some way - having some ability to cope if a pin isn't available. So far that's always been done with `!= NC` tests in the language, which covers both run-time and compile-time (can't do it in the preprocessor, cos pin values are enums). And that's also been used for different layers of "not available". Eg a config file might have explicitly set an `SPI_CS` to `NC`, or it might be `D1` and that in turn might be `NC`. 

And then SPI.cpp  avoids wiggling it if it is `NC`. It uses `if (_sw_ssel.is_connected())` which means "not constructed with `NC`" - the API implies that `NC` something you can see and test for. It's functionally like `nullptr`, really.

Not compiling would be fine if there was some way to `#ifdef` around it, for people who did want to be adaptive.

     #ifdef FF_ARDUINO_HAVE_D0
     DigitalOut cs(D3);
     #else
     MyDummyDigitalOut cs;    // This class would be a good idea, right?
     #endif

That sort of thing could actually lead to smaller code, by excluding the cs code at compile time. `if (PIN != NC)` directly like that can be compile-time excluded, but more general code like `DigitalOut(PIN)` won't be.

But then it doesn't fit the pattern of having generic `Driver(PinName, PinName)` classes (or the static pinmap equivalent). We treat pin selections as constructor data, not template parameters. We need to be able pass the null pins to the drivers.

With enough template brain-power, I could imagine

    DigitalOut<D3> cs;

and I think D3 not being defined is syntactically legal, and could select a dummy specialisation... But that's a whole other project.

Reply:
> Most uses of `NC` lead to either an MBED_ERROR or MBED_ASSERT
@0xc0170 How can we confirm this?

Reply:
Only to browse the implementation. It was a practice as mentioned above. You either allow NC in the target implementation (most of the cases are there in HAL implementation) or disallow (error or assert in this cases). Does that answer the question?

Reply:
@0xc0170 @kjbracey-arm 
By defining a pin in the PinName enum as NC, the application programs are enabled to test for "no valid connection" (or "not available"). Shouldn't such tests be extended to HAL for all targets to achieve actual code portability? Just wanted to make sure I am understanding this right.